js/app/downloader.js

```diff
@@ -114,44 +72,45 @@ const FileDownloader = new Lang.Class({
                     return status >= 200 && status < 300;
                 }
 
-                let body = request.send_finish(result);
                 let msg = request.get_message();
                 let status = msg.status_code;
 
-                if (status === Soup.Status.REQUESTED_RANGE_NOT_SATISFIABLE) {
-                    // The server is telling us we already have the full file range.
-                    finish_download(keep_new_file);
-                } else if (!soup_status_is_successful(status)) {
+                // The server is telling us we already have the full file range.
+                if (status === Soup.Status.REQUESTED_RANGE_NOT_SATISFIABLE)
+                    return true;
+
+                if (!soup_status_is_successful(status))
                     throw new Error("Server returned code " + status + " while fetching " + uri.to_string(true));
+
+                let out_stream;
+                if (status === Soup.Status.PARTIAL_CONTENT) {
+                    out_stream = tmp_file.append_to(Gio.FileCreateFlags.NONE, cancellable);
                 } else {
-                    let out_stream;
-                    if (status === Soup.Status.PARTIAL_CONTENT) {
-                        out_stream = tmp_file.append_to(Gio.FileCreateFlags.NONE, cancellable);
-                    } else {
-                        out_stream = tmp_file.replace(null, false, Gio.FileCreateFlags.NONE, cancellable);
-                    }
-
-                    out_stream.splice_async(body, Gio.OutputStreamSpliceFlags.NONE,
-                                            GLib.PRIORITY_DEFAULT, cancellable, task.catch_callback_errors((source, res) => {
-                                                out_stream.splice_finish(res);
-                                                out_stream.close(cancellable);
-                                                finish_download(keep_new_file);
-                                            }));
+                    out_stream = tmp_file.replace(null, false, Gio.FileCreateFlags.NONE, cancellable);
                 }
-            }));
-        });
 
-        return task;
-    },
+                return out_stream.splice_promise(body, Gio.OutputStreamSpliceFlags.NONE, GLib.PRIORITY_DEFAULT, cancellable).then(() => {
+                    out_stream.close(cancellable);
+                });
+            }).then(() => {
+                // If we have a checksum, verify it.
+                let csum = file_download_request.csum;
 
-    download_file_finish: function (task) {
-        return task.finish();
+                if (!csum)
+                    return true;
+                // XXX: If the file didn't checksum correctly, what should we do? Trash and re-download?
+                return this.verify_checksum_promise(tmp_file, csum, cancellable);
+            }).then(() => {
+                tmp_file.move(out_file, Gio.FileCopyFlags.OVERWRITE, cancellable, null);
```

Comment:
Consider wrapping the `move()` call in a promise as well, now that that's trivial